models/device.go

```diff
@@ -26,45 +26,48 @@ import (
  */
 type Device struct {
 	DescribedObject
-	Id             string         `json:"id"`
-	Name           string         `json:"name"`           // Unique name for identifying a device
-	AdminState     AdminState     `json:"adminState"`     // Admin state (locked/unlocked)
-	OperatingState OperatingState `json:"operatingState"` // Operating state (enabled/disabled)
-	Addressable    Addressable    `json:"addressable"`    // Addressable for the device - stores information about it's address
-	LastConnected  int64          `json:"lastConnected"`  // Time (milliseconds) that the device last provided any feedback or responded to any request
-	LastReported   int64          `json:"lastReported"`   // Time (milliseconds) that the device reported data to the core microservice
-	Labels         []string       `json:"labels"`         // Other labels applied to the device to help with searching
-	Location       interface{}    `json:"location"`       // Device service specific location (interface{} is an empty interface so it can be anything)
-	Service        DeviceService  `json:"service"`        // Associated Device Service - One per device
-	Profile        DeviceProfile  `json:"profile"`        // Associated Device Profile - Describes the device
+	Id             string                       `json:"id"`
+	Name           string                       `json:"name"`           // Unique name for identifying a device
+	AdminState     AdminState                   `json:"adminState"`     // Admin state (locked/unlocked)
+	OperatingState OperatingState               `json:"operatingState"` // Operating state (enabled/disabled)
+	Protocols      map[string]map[string]string `json:"protocols"`      // A map of supported protocols for the given device
```

Comment:
Can we typedef ProtocolProperties as map[string]string, and make the type for Protocols be map[string]ProtocolProperties

Reply:
No because as discussed on the call yesterday, this would require ProtocolProperties to be a struct type -- which is the same as we have right now with Addressable. This is considered too restrictive for the desired solution since a struct has a fixed list of properties that aren't flexible enough to support all cases. If it were, we'd make a map[string]Addressable and call it a day.